### ISSUE TITLE
irssi closes with "Lost terminal" signal

### ISSUE THREAD

**Issue by serpent7776** [irssi/irssi#1177]
Recently I've hitting by "Lost terminal" signal causing irssi to exit. The terminal that launches irssi closes itself too even though I'm not using `exec`. It happens randomly when switching windows.
Signal raised here: https://github.com/irssi/irssi/blob/1.2.2/src/fe-text/term-terminfo.c#L742
most probably by `EOF` while reading from `current_term->in`, which seems valid (i.e. has nonzero value).
Doesn't seem to be issue with irssi, but not sure what's the root cause. Is that the issue with the terminal itself?

Full backtrace:
```
* thread #1, name = 'irssi', stop reason = signal SIGHUP
  * frame #0: 0x0000000000474b50 irssi`sig_server_quit(server=0x5bb74175154378b2, msg=0x0000000000000000) at fe-server.c:405
    frame #1: 0x00000000004ee3c3 irssi`signal_emit_real(rec=0x0000000803879c00, params=2, va=0x00007fffffffe4c0, first_hook=0x0000000803887d00) at signals.c:242:3
    frame #2: 0x00000000004ee1cd irssi`signal_emit(signal="server quit", params=2) at signals.c:286:3
    frame #3: 0x00000000004f4004 irssi`cmd_disconnect(data="* Lost terminal", server=0x0000000803932000) at chat-commands.c:304:2
    frame #4: 0x00000000004f410a irssi`cmd_quit(data="Lost terminal") at chat-commands.c:327:3
    frame #5: 0x00000000004ee3c3 irssi`signal_emit_real(rec=0x0000000803879660, params=1, va=0x00007fffffffe800, first_hook=0x000000080385bdb0) at signals.c:242:3
    frame #6: 0x00000000004ee1cd irssi`signal_emit(signal="command quit", params=1) at signals.c:286:3
    frame #7: 0x000000000043ac9c irssi`term_gets(buffer=0x0000000803823d60, line_count=0x0000000000751994) at term-terminfo.c:742:3
    frame #8: 0x0000000000423fcc irssi`sig_input at gui-readline.c:838:3
    frame #9: 0x00000000004d771a irssi`irssi_io_invoke(source=0x0000000803899880, condition=17, data=0x00000008038bdaa0) at misc.c:51:3
    frame #10: 0x0000000801a49837 libglib-2.0.so.0`g_main_context_dispatch + 327
    frame #11: 0x0000000801a49bfa libglib-2.0.so.0`___lldb_unnamed_symbol117$$libglib-2.0.so.0 + 538
    frame #12: 0x0000000801a49cb4 libglib-2.0.so.0`g_main_context_iteration + 100
    frame #13: 0x0000000000444de2 irssi`main(argc=1, argv=0x00007fffffffeab8) at irssi.c:351:3
    frame #14: 0x000000000041d57d irssi`_start + 141
```

**Comment by dequis** [comment]
What terminal?

**Comment by serpent7776** [comment]
st v0.8.2
https://st.suckless.org/

**Comment by dequis** [comment]
So it seems to send SIGHUP to the containing shell on `ttyhangup()` which itself is called from `cmessage` when it receives a WM_DELETE_WINDOW ClientMessage, that is, your window manager sending a "close window" message. Normal stuff. Maybe set a breakpoint on st's ttyhangup to confirm this.

As an aside, I'm somewhat confused at how the irssi traceback says SIGHUP but it's not raised through a signal handler, this doesn't match my understanding of signals. If anyone gets that part I'd love to know what's going on there.

**Comment by dequis** [comment]
Did you get to look into this?

(Personally I always use tmux inside st, as they intentionally don't implement things like scrollback because tmux can do it. That should also sort your hangups)

**Comment by serpent7776** [comment]
I did not receive any sighup since I opened the issue. I did upgrade some of the packages (like my window manager) and rebuilt `st` with debug symbols. Maybe it got magically fixed itself. I'll investigate some more if I get the next sighup.
I don't like running irssi inside tmux as irssi already has its own scrollback, besides I'm using `st` with scrollback patch. Maybe I should consider `dtach` tool.
